staging/src/k8s.io/client-go/examples/third-party-resources/watcher.go

```diff
@@ -0,0 +1,148 @@
+	// in case of resource-generated evxents.
+	if !cache.WaitForCacheSync(ctx.Done(), exampleInformer.HasSynced) {
+		return errors.New("Wait for Example Informer was cancelled")
+	}
+
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func watchExamples(ctx context.Context, exampleClient cache.Getter, exampleScheme *runtime.Scheme, handler cache.ResourceEventHandler) (cache.Controller, error) {
+	parameterCodec := runtime.NewParameterCodec(exampleScheme)
+
+	source := newListWatchFromClient(
+		exampleClient,
+		ExampleResourcePath,
+		api.NamespaceAll,
+		fields.Everything(),
+		parameterCodec)
+
+	store, controller := cache.NewInformer(
+		source,
+
+		// The object type.
+		&Example{},
+
+		// resyncPeriod
+		// Every resyncPeriod, all resources in the cache will retrigger events.
+		// Set to 0 to disable the resync.
+		0,
+
+		// Your custom resource event handlers.
+		handler)
+
+	// store can be used to List and Get
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	for _, obj := range store.List() {
+		example := obj.(*Example)
+
+		// This will likely be empty the first run, but may not
+		fmt.Printf("Existing example: %#v\n", example)
+	}
+
+	go controller.Run(ctx.Done())
+
+	return controller, nil
+}
+
+// See the issue comment: https://github.com/kubernetes/kubernetes/issues/16376#issuecomment-272167794
+// newListWatchFromClient is a copy of cache.NewListWatchFromClient() method with custom codec
+// Cannot use cache.NewListWatchFromClient() because it uses global api.ParameterCodec which uses global
+// api.Scheme which does not know about custom types (Example in our case) group/version.
+func newListWatchFromClient(c cache.Getter, resource string, namespace string, fieldSelector fields.Selector, paramCodec runtime.ParameterCodec) *cache.ListWatch {
+	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
+		return c.Get().
+			Namespace(namespace).
+			Resource(resource).
+			VersionedParams(&options, paramCodec).
+			FieldsSelectorParam(fieldSelector).
+			Do().
+			Get()
+	}
+	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
+		return c.Get().
+			Prefix("watch").
+			Namespace(namespace).
+			Resource(resource).
+			VersionedParams(&options, paramCodec).
+			FieldsSelectorParam(fieldSelector).
+			Watch()
+	}
+	return &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// ExampleEventHandler can handle events for Example resource
+type ExampleEventHandler struct {
+}
+
+func (h *ExampleEventHandler) OnAdd(obj interface{}) {
+	example := obj.(*Example)
+	log.Printf("[WATCH] OnAdd %s\n", example.Metadata.SelfLink)
+}
+
+func (h *ExampleEventHandler) OnUpdate(oldObj, newObj interface{}) {
+	oldExample := oldObj.(*Example)
+	newExample := newObj.(*Example)
+	log.Printf("[WATCH] OnUpdate oldObj: %s\n", oldExample.Metadata.SelfLink)
+	log.Printf("[WATCH] OnUpdate newObj: %s\n", newExample.Metadata.SelfLink)
```

Comment:
Nit: `log.Print*()` does not need `\n` at the end, it will be added automatically. Also some places use `fmt.Print()` - this is inconsistent. Should probably do one or another.